# Diplodatos Kaggle Competition

We present this peace of code to create the baseline for the competition, and as an example of how to deal with these kind of problems. The main goals are that you:

1. Learn
1. Try different models and see which one fits the best the given data
1. Get a higher score than the given one in the current baseline example
1. Try to get the highest score in the class :)

In [12]:
# Import the required packages
import os

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [13]:
# load the given labels
breed = pd.read_csv('../data/breed_labels.csv')
color = pd.read_csv('../data/color_labels.csv')
state = pd.read_csv('../data/state_labels.csv')

Now we take a look at the labels, just to understand what these are

In [14]:
breed.head()

BreedID  Type         BreedName
0        1     1     Affenpinscher
1        2     1      Afghan Hound
2        3     1  Airedale Terrier
3        4     1            Akbash
4        5     1             Akita

In [15]:
color.head()

ColorID ColorName
0        1     Black
1        2     Brown
2        3    Golden
3        4    Yellow
4        5     Cream

In [16]:
state

StateID        StateName
0     41336            Johor
1     41325            Kedah
2     41367         Kelantan
3     41401     Kuala Lumpur
4     41415           Labuan
5     41324           Melaka
6     41332  Negeri Sembilan
7     41335           Pahang
8     41330            Perak
9     41380           Perlis
10    41327     Pulau Pinang
11    41345            Sabah
12    41342          Sarawak
13    41326         Selangor
14    41361       Terengganu

And now we are ready to deal with the *original* dataset...

In [17]:
original_df = pd.read_csv('../data/train.csv')

In [18]:
original_df.columns

Index(['Type', 'Age', 'Breed1', 'Breed2', 'Gender', 'Color1', 'Color2',
       'Color3', 'MaturitySize', 'FurLength', 'Vaccinated', 'Dewormed',
       'Sterilized', 'Health', 'Quantity', 'Fee', 'State', 'Description',
       'AdoptionSpeed', 'PID'],
      dtype='object')

In [19]:
original_df.describe()

Type           Age        Breed1        Breed2        Gender  \
count  10582.000000  10582.000000  10582.000000  10582.000000  10582.000000   
mean       1.454734     10.520412    265.469854     74.388868      1.779059   
std        0.497970     18.374027     60.121490    123.434010      0.684763   
min        1.000000      0.000000      0.000000      0.000000      1.000000   
25%        1.000000      2.000000    265.000000      0.000000      1.000000   
50%        1.000000      3.000000    266.000000      0.000000      2.000000   
75%        2.000000     12.000000    307.000000    188.000000      2.000000   
max        2.000000    255.000000    307.000000    307.000000      3.000000   

             Color1        Color2        Color3  MaturitySize     FurLength  \
count  10582.000000  10582.000000  10582.000000  10582.000000  10582.000000   
mean       2.230675      3.236912      1.856738      1.860518      1.460971   
std        1.743985      2.748595      2.974465      0.547535      0.593843   
min        1.000000      0.000000      0.000000      1.000000      1.000000   
25%        1.000000      0.000000      0.000000      2.000000      1.000000   
50%        2.000000      2.000000      0.000000      2.000000      1.000000   
75%        3.000000      6.000000      5.000000      2.000000      2.000000   
max        7.000000      7.000000      7.000000      4.000000      3.000000   

         Vaccinated      Dewormed    Sterilized        Health      Quantity  \
count  10582.000000  10582.000000  10582.000000  10582.000000  10582.000000   
mean       1.729730      1.566528      1.912115      1.036666      1.584011   
std        0.670791      0.701482      0.564041      0.198228      1.488348   
min        1.000000      1.000000      1.000000      1.000000      1.000000   
25%        1.000000      1.000000      2.000000      1.000000      1.000000   
50%        2.000000      1.000000      2.000000      1.000000      1.000000   
75%        2.000000      2.000000      2.000000      1.000000      1.000000   
max        3.000000      3.000000      3.000000      3.000000     20.000000   

                Fee         State  AdoptionSpeed           PID  
count  10582.000000  10582.000000   10582.000000  10582.000000  
mean      20.809960  41345.994613       2.518900   7477.025799  
std       78.397243     32.409109       1.176018   4310.921553  
min        0.000000  41324.000000       0.000000      0.000000  
25%        0.000000  41326.000000       2.000000   3768.250000  
50%        0.000000  41326.000000       2.000000   7473.500000  
75%        0.000000  41401.000000       4.000000  11200.750000  
max     3000.000000  41415.000000       4.000000  14992.000000

Create a function to transform the datasets. This is done by means of a function so that the transformations are the same for the training and testing datasets... We replace the encodings just to make it easy to "visualize" the data

In [20]:
def transform_data(train_data_fname, test_data_fname):
    def transform_columns(df):
        df = df.drop(["Description"], axis=1)
        df.Type = df.Type.replace({1: 'Dog', 2: 'Cat'})
        df.Gender = df.Gender.replace({1:'Male', 2:'Female', 3:'Mixed'})
        df.MaturitySize = df.MaturitySize.replace({1:'S', 2:'M', 3:'L', 4:'XL', 0:'N/A'})
        df.FurLength = df.FurLength.replace({1:'S', 2:'M', 3:'L', 0:'N/A'})
        df.Vaccinated = df.Vaccinated.replace({1:'T', 2:'N', 3:'N/A'})
        df.Dewormed = df.Dewormed.replace({1:'T', 2:'F', 3:'N/A'})
        df.Sterilized = df.Sterilized.replace({1:'T', 2:'F', 3:'N/A'})
        df.Health = df.Health.replace({1:'Healthy', 2: 'MinorInjury', 3:'SeriousInjury', 0: 'N/A'})
        df.Color1 = df.Color1.replace(dict(list(zip(color.ColorID, color.ColorName)) + [(0, "N/A")]))
        df.Color2 = df.Color2.replace(dict(list(zip(color.ColorID, color.ColorName)) + [(0, "N/A")]))
        df.Color3 = df.Color3.replace(dict(list(zip(color.ColorID, color.ColorName)) + [(0, "N/A")]))
        df.Breed1 = df.Breed1.replace(dict(list(zip(breed.BreedID, breed.BreedName)) + [(0, "N/A")]))
        df.Breed2 = df.Breed2.replace(dict(list(zip(breed.BreedID, breed.BreedName)) + [(0, "N/A")]))
        return df
    
    df_train = pd.read_csv(train_data_fname)
    df_train = transform_columns(df_train)
    df_test = pd.read_csv(test_data_fname)
    df_test = transform_columns(df_test)
    
    df = pd.concat([df_train, df_test], sort=True)

    # set dummy variables for everything
    # except from Age, Quantity, Fee
    df = pd.get_dummies(df)
    # get train and test back
    n = len(df_train)
    df_train = df.iloc[:n]
    df_test = df.iloc[n:]
    
    y = df_train['AdoptionSpeed']
    X = df_train.drop('AdoptionSpeed', axis=1)
    yy = None
    XX = df_test.drop('AdoptionSpeed', axis=1)

    return X, y, XX, yy

Load the data...

In [21]:
X, y, XX, yy = transform_data("../data/train.csv", "../data/test.csv")

Create the model and evaluate it

#### Probamos un Decision Tree. Los hiperparámetros que evaluamos, con Grid Search, son:
- criterio: gini y entropy
- min_samples_leaf [1...7]
- min_samples_split [2 ... 200]

In [83]:
# split training dataset into train and "validation" 
# (we won't be using validation set in this example, because of the cross-validation;
# but it couldn be useful for you depending on your approach)
from sklearn.model_selection import train_test_split
X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.3, random_state=42)

from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV

results = pd.DataFrame(columns=('clf', 'best_acc'))

from sklearn.tree import DecisionTreeClassifier as DT
tree_param = {'criterion':('gini', 'entropy'), 'min_samples_leaf':(1, 2, 5,7),
              'min_samples_split':(2, 3, 5, 10,25, 50, 100,200)}
             
tree = DT(random_state=42)
tree_clf = GridSearchCV(tree, tree_param, scoring='accuracy', cv=5, iid=False)
tree_clf.fit(X_train.drop(["PID"], axis=1), y_train)
best_tree_clf = tree_clf.best_estimator_
#score general de accuracy
print('Best Decision Tree accuracy: ', tree_clf.best_score_)
print(best_tree_clf)
results = results.append({'clf': best_tree_clf, 'best_acc': tree_clf.best_score_}, ignore_index=True)

print('The best classifier so far is: ')
print(results.loc[results['best_acc'].idxmax()]['clf'])

Best Decision Tree accuracy:  0.37234878199372645
DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=7, min_samples_split=200,
            min_weight_fraction_leaf=0.0, presort=False, random_state=42,
            splitter='best')
The best classifier so far is: 
DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=7, min_samples_split=200,
            min_weight_fraction_leaf=0.0, presort=False, random_state=42,
            splitter='best')


#### Los mejores hiperparámetros que selecciona grid search son:
- criterio : entropy
- min_samples_leaf : 7
- min_samples_split : 200

In [84]:
# vuelvo a fitear el modelo con los mejores hiperparámetros, pero usando todo el entrenamiento
bb_clf = DT(**tree_clf.best_params_)
bb_clf.fit(X_train.drop(["PID"], axis=1), y_train)
# uso el modelo para predecir sobre el x_train
y_train_pred = bb_clf.predict(X_train.drop(["PID"], axis=1))
# uso el modelo para predecir con los datos de validación (x_valid)
y_valid_pred = bb_clf.predict(X_valid.drop(["PID"], axis=1))

In [85]:
# métricas de accuracy para entrenamiento y validación para chequear calidad del modelo y posible overfitting
print (accuracy_score(y_valid_pred, y_valid))
print (accuracy_score(y_train, y_train_pred))

0.3700787401574803
0.41177264749561227


#### No hay demasiada diferencia entre entrenamiento y validación, lo que sugiere poco overfitting. Da 0.37 con la validación

In [88]:
#yy = results.clf.iloc[0].predict(XX.drop(["PID"], axis=1))
yy = bb_clf.predict(XX.drop(["PID"], axis=1))
yy = yy.astype(np.int)
submission = pd.DataFrame(list(zip(XX.PID, yy)), columns=["PID", "AdoptionSpeed"])
submission.to_csv("../data/submission_mejor.csv", header=True, index=False)

#### Luego probamos con más opciones para min_samples_leaf y min_samples_split, porque ambos daban los mejores hiperparámetros en el máximo del rango que probamos para ambos hiperparámetros (200 y 7, respectivamente). También agregamos class_weight para probar balanceada, además del None que es el valor por defecto. 

In [89]:
# split training dataset into train and "validation" 
# (we won't be using validation set in this example, because of the cross-validation;
# but it couldn be useful for you depending on your approach)
from sklearn.model_selection import train_test_split
X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.3, random_state=42)

from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV


results = pd.DataFrame(columns=('clf', 'best_acc'))

from sklearn.tree import DecisionTreeClassifier as DT
tree_param = {'criterion':('entropy',), 'min_samples_leaf':(7,10, 12, 15),
              'min_samples_split':(100,200, 300, 400), 'max_leaf_nodes':(None,),
             'class_weight':('balanced',None) }

tree = DT(random_state=42)
tree_clf = GridSearchCV(tree, tree_param, scoring='accuracy', cv=5, iid=False)
tree_clf.fit(X_train.drop(["PID"], axis=1), y_train)
best_tree_clf = tree_clf.best_estimator_
#score general de accuracy
print('Best Decision Tree accuracy: ', tree_clf.best_score_)
print(best_tree_clf)
results = results.append({'clf': best_tree_clf, 'best_acc': tree_clf.best_score_}, ignore_index=True)

print('The best classifier so far is: ')
print(results.loc[results['best_acc'].idxmax()]['clf'])

Best Decision Tree accuracy:  0.374508938839338
DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=12, min_samples_split=200,
            min_weight_fraction_leaf=0.0, presort=False, random_state=42,
            splitter='best')
The best classifier so far is: 
DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=12, min_samples_split=200,
            min_weight_fraction_leaf=0.0, presort=False, random_state=42,
            splitter='best')


#### El min_samples_split óptimo sigue siendo 200. Para min_samples_leaf, el mejor hiperparámetro da 12 (en vez de 7 como en el caso anterior). Para class_weight, sigue optimizándose con igual peso todas las clases.

In [90]:
# fiteo con todo el entrenamiento
bb_clf = DT(**tree_clf.best_params_)
bb_clf.fit(X_train.drop(["PID"], axis=1), y_train)
y_train_pred = bb_clf.predict(X_train.drop(["PID"], axis=1))
y_valid_pred = bb_clf.predict(X_valid.drop(["PID"], axis=1))

In [91]:
# métricas del modelo
print (accuracy_score(y_valid_pred, y_valid))
print (accuracy_score(y_train, y_train_pred))

0.3640944881889764
0.4137977588767382


#### Da bastante similar al caso anterior. Ligeramente inferior para el test (0.364 en vez de 0.37). Nos quedamos con el modelo anterior con min_samples_leaf == 7

**And finally**, we predict the unknown label for the testing set

In [92]:
X.shape, XX.shape

((10582, 360), (4411, 360))

In [93]:
#yy = results.clf.iloc[0].predict(XX.drop(["PID"], axis=1))
yy = bb_clf.predict(XX.drop(["PID"], axis=1))
yy = yy.astype(np.int)
submission = pd.DataFrame(list(zip(XX.PID, yy)), columns=["PID", "AdoptionSpeed"])
submission.to_csv("../data/submission_pr2.csv", header=True, index=False)

The last thing we do is generating a file that should be *submitted* on kaggle

#### Probamos otros modelos pero tampoco obtuvimos mejora significativa respecto a este. Todos dan en el mismo orden. Probamos random forest, un Knn y SGD. Todos los resultados están alrededor de 0.35-0.36